lib/saluki-env/src/workload/stores/external_data.rs

```diff
@@ -0,0 +1,266 @@
+use std::{num::NonZeroUsize, sync::Arc};
+
+use arc_swap::ArcSwap;
+use memory_accounting::{MemoryBounds, MemoryBoundsBuilder};
+use tracing::debug;
+
+use crate::{
+    prelude::*,
+    workload::{
+        aggregator::MetadataStore,
+        external_data::{ExternalData, ExternalDataRef},
+        EntityId, MetadataAction, MetadataOperation,
+    },
+};
+
+/// A store for External Data entity mappings.
+pub struct ExternalDataStore {
+    snapshot: Arc<ArcSwap<ExternalDataSnapshot>>,
+
+    entity_limit: NonZeroUsize,
+    active_entities: FastHashSet<EntityId>,
+
+    forward_mappings: FastIndexMap<ExternalData, EntityId>,
+    reverse_mappings: FastIndexMap<EntityId, ExternalData>,
+}
+
+impl ExternalDataStore {
+    /// Creates a new `ExternalDataStore` with the given entity limit.
+    ///
+    /// The entity limit is the maximum number of unique entities that can be stored. Once the limit is reached, new
+    /// entities will not be added to the store.
+    pub fn with_entity_limit(entity_limit: NonZeroUsize) -> Self {
+        Self {
+            snapshot: Arc::new(ArcSwap::new(Arc::new(ExternalDataSnapshot::default()))),
+            entity_limit,
+            active_entities: FastHashSet::default(),
+            forward_mappings: FastIndexMap::default(),
+            reverse_mappings: FastIndexMap::default(),
+        }
+    }
+
+    /// Returns the maximum number of unique entities that can be tracked by the store at any given time.
+    pub fn entity_limit(&self) -> usize {
+        self.entity_limit.get()
+    }
+
+    fn track_entity(&mut self, entity_id: &EntityId) -> bool {
+        if self.active_entities.contains(entity_id) {
+            return true;
+        }
+
+        if self.active_entities.len() >= self.entity_limit() {
+            return false;
+        }
+
+        let _ = self.active_entities.insert(entity_id.clone());
+        true
+    }
+
+    fn add_mapping(&mut self, external_data: ExternalData, entity_id: EntityId) {
+        if !self.track_entity(&entity_id) {
+            // TODO: Emit a warning log and/or a metric here.
```

Comment:
TODO intentionally left?

Reply:
Yeah, good callout: added intentional to not forget, but not meant to be left.

Updated this, and also updated the equivalent ones in `TagStore` as well, for consistency.